### DISCUSSION TITLE
Expansion to Buckets on Cylindrical geometries

### DISCUSSION THREAD

**Discussion by robbietuk** [UCL/STIR#1388] [Ideas]
Primarily an extension of issue https://github.com/UCL/STIR/issues/1374 and PR https://github.com/UCL/STIR/pull/1375. The motivation for this was detailed [here](https://github.com/UCL/STIR/pull/1375#issuecomment-1939290595) with an initial plan laid out. This issue serves as an extension of that comment.

STIR does not support multiple axial buckets of blocks of crystals on a ring. https://github.com/UCL/STIR/pull/1375 disabled the ability to use multiple axial buckets due to a bug in an assumptive implementation. 

I require the use of `gaps_between blocks != gaps_between_buckets` therefore I am planning the implementation of `BucketsOnCylindrical` using the `BlocksOnCylindrical` current frame work. I do not plan to change the name, just the implementation.

Below are some of the main points worth discussion.

_____________
What do are a set of parallel buckets called?
----

_I will refer to them as `modules` in this issue but the name can change via a refactor._

_Modules_ are rotated by angle `alpha` around the z axis of the scanner. It is important a suitable name is given to the collection of buckets

Some other ideas:

- In GATE these are called `rsectors` for CylindricalPET (see [GATE example `.mac` file](https://github.com/UCL/STIR-GATE-Connection/blob/master/ExampleScanners/D690/geometry_D690.mac)) 
https://github.com/UCL/STIR/blob/37f799fd1bd51a19eeb0c95f9128816a3bd9959a/documentation/STIR-UsersGuide.tex#L779-L797
STIR could adopt `rsectors`, `sectors` or `sections`
- Alternatively, these can be refereed to as a `modules`. This would might make the translation between STIR and GATE even more annoying but it is a good fit.
- @KrisThielemans I don't know exactly what other manufacturers refer to them as.

Suggestions welcome

_____________
Transaxial Buckets are maybe not supported?
-------
>Looking at the [current definition](https://github.com/UCL/STIR/blob/ee307c721cae562ab163aee06dbab8407d861414/src/buildblock/GeometryBlocksOnCylindrical.cxx#L111-L129),the alpha definition is in terms of buckets. So it will put all blocks in a bucket next/parallel to eachother. That looks good to me. So, I think that currently BlocksOnCylindrical does support multiple blocks per bucket in transaxial direction. However, clearly in axial direction the bucket location is completely ignored.

https://github.com/UCL/STIR/pull/1375#issuecomment-1935082119.  I am not sure this is the case upon further review. 
https://github.com/UCL/STIR/blob/37f799fd1bd51a19eeb0c95f9128816a3bd9959a/src/buildblock/GeometryBlocksOnCylindrical.cxx#L128-L129
It seems `trans_bucket_num` is only used in the computation of `alpha` and in the `tangential_coord` (similar to `axial_bucket_num` in calculation of `axial_coord`). The `alpha` is used to determine the `rotation_matrix`, i.e. `alpha` is an angle around the z axis.

I understand this as no two buckets can be parallel.

In this case, additional variables is needed throughout: `num_axial_buckets_per_module` and `num_transaxial_buckets_per_module`. 

- These variables inevitably need to be everywhere but will default to `1`. 
- `GeometryBlocksOnCylindrical::build_crystal_maps` loops will need to be extended.
- More methods added to `get_num_modules()`, `get_num_axial_crystals_per_module()` and `get_num_transaxial_crystals_per_module()`

_____________
Spacing between buckets (Axial & Transaxial)
------
> Furthermore, for most scanners it would make more sense to have a larger axial gap between buckets than between blocks. We currently don't have the variables for that.

https://github.com/UCL/STIR/pull/1375#issuecomment-1935082119 . Not too much to add. The value should probably default to `num_axial_blocks_per_bucket*get_axial_block_spacing()`, otherwise a user defined variable > 0.

_________
More tests
-------

https://github.com/UCL/STIR/pull/1375 adds a test requiring the axial crystals position to increase monotonically with the axial crystal coordinate. 

Additional tests should be added for the transaxial. As an idea, the angle sequential coordinate vectors in the detector map should never be greater than 90 degrees, i.e. `cos(A -> B, B->C) >= 0`. _Except maybe in the instance of only two modules_.

_____________
Virtual Crystals
------
https://github.com/UCL/STIR/pull/776 by @NikEfth tries to add virtual crystals. I am not 100% sure on the status of this and neither do I want to delve into it but I assume any issues encountered with adding virtual crystals to `BucketsOnCylindrical` would also be present in the `BlocksOnCylindrical`, so no significant change there. 

_____________
Data Correction
------
Will there be any other issues with data correction methods etc. that we have not taken into account? e.g. scatter correction and normalization? I expect the same issues, maybe further complicated by this proposal.

@KrisThielemans @markus-jehl @danieldeidda

**Comment by robbietuk** [comment]
Notes for myself.

It is currently assumed that a scanner is defined from two parts. Consider `Scanner::set_params()` method. A "module", _as dubbed above_, is defined from the following transaxial variables:
https://github.com/UCL/STIR/blob/79a1e83de1141ad42db61f5bf2de7a2b7827e8f5/src/buildblock/Scanner.cxx#L1428-L1431

The number of these "modules", incrementally rotated by angle `alpha` about the z axis, is derived from this transaxial bucket information and the `num_detectors_per_ring_v` parameter. _I am not sure there is a check on this..._

 The axial information is derived from the above axial variables in the code quote and the `num_rings_v`. 

To expand to BucketsOnCylindrical, "modules" need to be derived from the bucket per module details too. This will require an expansion in the scanner definition to add `num_axial_buckets_per_module_v` and `num_transaxial_buckets_per_module_v`. Checks should be run on all of these.

_______
I should also mention that in https://github.com/UCL/STIR/pull/1375#issuecomment-1934876335 I suggested replacing `num_rings` (and by extension `num_detectors_per_ring`):
> Transition away from the use of num_rings to define the geometry and implement entirely from num_crystals_per_block, num_blocks_per_bucket and num_buckets_per_(module?) (in each dimension)
>    - _I personally perfer this method as it is the simplest for the user as they no longer need to precalculate the num_rings, STIR does it for them._

This is method is similar to how GATE handles PET geometries and it doesn't require the user to pre-calculate `num_rings` and `num_detectors_per_ring` in the scanner definition. It would also remove the potential issues with getting the value wrong... However, that depreciates functionality and is a rather large change. I propose using the current tooling and then make the switch later if desired, maybe STIR v7?

**Comment by KrisThielemans** [comment]
# What are a set of parallel buckets called?

Do you mean "What are a set of parallel blocks called"? Maybe not... 

I'm still confused if you want to have a difference between "buckets" and "modules".

These days, PET scanners can have a whole hierarchy of things. I know of at least
- crystals (potentially with layers)
- mini-blocks
- blocks
- a collection of blocks (usually (always?) parallel) handled by the same electronics (relevant for dead-time)
- a collection of parallel blocks with equal spacing

Other notes:
- For all scanners I know, the last 2 are the same, but obviously there is no strict need for this.
- STIR doesn't encode mini-blocks, but they are not related to geometry (but could be to dead-time)
- STIR doc:
   - https://github.com/UCL/STIR/blob/79a1e83de1141ad42db61f5bf2de7a2b7827e8f5/documentation/STIR-glossary.tex#L59-L66
   - https://github.com/UCL/STIR/blob/79a1e83de1141ad42db61f5bf2de7a2b7827e8f5/src/include/stir/Scanner.h#L66-L85
- Siemens uses "buckets" in their interfile headers, but they have been told to switched internally to "modules". GE uses "modules"

Given all this, I think that at present it would be enough to stick to 3 levels as we currently do, and the highest level can be called buckets or modules, but they mean the same thing in STIR. If this is not enough for your scanner, we need to do more thinking (and code modification). If you can, a description with enough detail of what you need would be necessary.

# Transaxial Buckets are maybe not supported?

> I understand this as no two buckets can be parallel.

That's correct for transaxial buckets (obviously, in axial direction, they still can be)

So, do you need more?

**Reply by robbietuk** [reply]
The geometry I am trying to model is the following :
- `n` x `n` crystals in a block
- `m` x `m` blocks in a bucket
- `i` axial buckets.
where `n>1`, `m>1` and `i>1`. Currently `i` is limited to 1 in STIR. This with `i>1`, the scanner would have a set of parallel buckets, which is what I have refereed to as a module. The reason I cannot restructure to put all axial blocks into the same axial bucket this is dead time and the gaps/pitch between buckets is not the same as the gaps between blocks.

Note, the extension I am proposing would also allow `>1` transaxial buckets.

**Reply by robbietuk** [reply]
> >  I understand this as no two buckets can be parallel.
>That's correct for transaxial buckets (obviously, in axial direction, they still can be)
> So, do you need more?

Perhaps not... I might have got ahead of myself with the `num_axial_buckets_per_module_v` and `num_transaxial_buckets_per_module_v` additional variable requirements.

**Reply by KrisThielemans** [reply]
So, I see no need for a module level then. We just say there are i x m x n rings, n crystals per axial block, m blocks per axial bucket.   We need a "bucket-spacing` capability. done.

> Note, the extension I am proposing would also allow >1 transaxial buckets.

now you've confused me again. I think the current code does what you need already on the transaxial front.

**Reply by robbietuk** [reply]
> So, I see no need for a module level then. We just say there are i x m x n rings, n crystals per axial block, m blocks per axial bucket. We need a "bucket-spacing` capability. done.

Yup! I over complicated it for myself.

> > Note, the extension I am proposing would also allow >1 transaxial buckets.
> 
> now you've confused me again. I think the current code does what you need already on the transaxial front.

I disagree.
https://github.com/UCL/STIR/blob/37f799fd1bd51a19eeb0c95f9128816a3bd9959a/src/buildblock/GeometryBlocksOnCylindrical.cxx#L128-L131
`alpha` is proportional to the `rotation_matrix` and increases with `trans_bucket_num`. Sequential transaxial buckets rotate.

**Reply by KrisThielemans** [reply]
> alpha is proportional to the rotation_matrix and increases with trans_bucket_num. Sequential transaxial buckets rotate.

`alpha` increases with `trans_bucket_num`, so all blocks in one trans_bucket are parallel. But isn't that what you want? (`alpha` is obviously not proportional to the matrix, but it must have been late).

However, see #1390

**Comment by robbietuk** [comment]
Is there an equation that explains https://github.com/UCL/STIR/blob/37f799fd1bd51a19eeb0c95f9128816a3bd9959a/src/buildblock/GeometryBlocksOnCylindrical.cxx#L89-L91

I believe it finds the distance from the first and last axial crystals and then divides by 2 but the exact methodology here is hard to understand. This should be refactored to make this clear.

**Reply by robbietuk** [reply]
I am pretty sure this can be replaced by

```
  float crystals_in_block_offset = (scanner.get_num_axial_crystals_per_block() - 1) * scanner.get_axial_crystal_spacing();
  float blocks_in_bucket_offset = (scanner.get_num_axial_blocks_per_bucket() - 1) * scanner.get_axial_block_spacing();
  float bucket_offset = (scanner.get_num_axial_buckets() - 1) * scanner.get_axial_bucket_spacing();
  return -0.5 * (crystals_in_block_offset + blocks_in_bucket_offset + bucket_offset);
```
The idea here is if the crystals in a block are centered, and the blocks in a bucket are centered, then the buckets are centered in the z axis, the scanner should be centered. Current tests look good.

**Comment by robbietuk** [comment]
I have an issue with the use of non-descriptive variable names and code readability in the BlocksOnCylindrical project. I don't understand half of the shorthand variables in https://github.com/UCL/STIR/blob/master/examples/python/plot_scanner_LORs.py and trying to comprehend how positions of detectors are calculated is challenging and likely prone to errors, see https://github.com/UCL/STIR/issues/1390 and the cause https://github.com/UCL/STIR/blob/37f799fd1bd51a19eeb0c95f9128816a3bd9959a/src/buildblock/GeometryBlocksOnCylindrical.cxx#L81-L98

**Comment by markus-jehl** [comment]
I agree with Robbie that in transaxial direction having the module concept would be required to support different spacings between blocks and buckets. We would actually require this with our scanner, but chose to spread out the overall sum of block and bucket spacing evenly on the module. Then we rely on the geometric normalisation to correct for the resulting small modelling errors. It would obviously be nice to be able to model any sort of geometry without resorting to hand-made crystal maps, but it could complicate the code a lot and lead to potential additional bugs that would be very hard to find.
